toolchain/lower/file_context.h

```diff
@@ -165,6 +184,69 @@ class FileContext {
 
   auto BuildVtable(const SemIR::Class& class_info) -> llvm::GlobalVariable*;
 
+  // Track which specifics were lowered for each generic. These are added one
+  // by one while lowering their definitions.
+  auto AddLoweredSpecificForGeneric(SemIR::GenericId generic_id,
+                                    SemIR::SpecificId specific_id) {
+    lowered_specifics_[generic_id.index].push_back(specific_id);
+  }
+
+  // Initialize a return a SpecificFunctionFingerprint* instance for specific.
+  // The internal of the fingerprint are populated during and after lowering
+  // the function body of that specific.
+  auto InitializeFingerprintForSpecific(SemIR::SpecificId specific_id)
+      -> SpecificFunctionFingerprint* {
+    if (!specific_id.has_value()) {
+      return nullptr;
+    }
+    return &lowered_specific_fingerprint_[specific_id.index];
+  }
+
+  // Entry point for coalescing equivalent specifics. Two function definitions,
+  // from the same generic, with different specific_ids are considered
+  // equivalent if, at the LLVM level, one can be replaced with the other, with
+  // no change in behavior. All LLVM types and instructions must be equivalent.
+  auto CoalesceEquivalentSpecifics() -> void;
+
+  // While coalescing specifics, compare the function types for two specifics.
```

Comment:
```suggestion
  // While coalescing specifics, returns whether the function types for two specifics are equivalent.
```

Note "compare" doesn't really work with "this function ..." (see see https://github.com/carbon-language/carbon-lang/pull/5314#discussion_r2092086566). But, maybe directly address what the return value is?